### ISSUE TITLE
stuck on 'setting up reader'

### ISSUE THREAD

**Issue by flawlessBadger** [iZettle/sdk-ios#274]
I am trying to process a card payment. I tried using the iZettleSDKSample app and my own react native implementation of the iZettleSDK. Both get stuck on the same step.
After calling the charge function I am able to login and connect an iZettle Reader 2. then I get stuck on a 'Setting up Reader, Please wait...' screen. The reader displays a 'Hello' message and every couple seconds flashes "Please wait".

This is the console log up to that point:
`2019-11-13 13:05:44.526328+0100 iZettle SDK Sample[392:13850] IAPDHasLaunched: kIAPAvailableNotification iapdAvailableState 0 -> 0
2019-11-13 13:05:44.526502+0100 iZettle SDK Sample[392:13850] IAP2DHasLaunched: kIAP2AvailableNotification iap2dAvailableState 0 -> 0
CoreData: annotation:  Failed to load optimized model at path '/var/containers/Bundle/Application/9E6650CA-EAA9-4FE8-BD88-096DF4AA70FB/iZettle SDK Sample.app/Frameworks/iZettlePayments.framework/Account.momd/Account 5.21.omo'
2019-11-13 13:05:44.572838+0100 iZettle SDK Sample[392:13850] [error] fault: One or more models in this application are using transformable properties with transformer names that are either unset, or set to NSKeyedUnarchiveFromDataTransformerName. Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead. At some point, Core Data will default to using "NSSecureUnarchiveFromData" when nil is specified, and transformable properties containing classes that do not support NSSecureCoding will become unreadable.
CoreData: fault: One or more models in this application are using transformable properties with transformer names that are either unset, or set to NSKeyedUnarchiveFromDataTransformerName. Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead. At some point, Core Data will default to using "NSSecureUnarchiveFromData" when nil is specified, and transformable properties containing classes that do not support NSSecureCoding will become unreadable.
2019-11-13 13:05:44.573029+0100 iZettle SDK Sample[392:13850] [error] CoreData: One or more models in this application are using transformable properties with transformer names that are either unset, or set to NSKeyedUnarchiveFromDataTransformerName. Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead. At some point, Core Data will default to using "NSSecureUnarchiveFromData" when nil is specified, and transformable properties containing classes that do not support NSSecureCoding will become unreadable.
CoreData: warning: Property 'baseURL' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'baseURLs' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'cardPaymentConfig' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'payload' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'sessionCacheJSON' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'tokensForDisplayedInformation' on Entity 'Account' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'payload' on Entity 'TransactionReceiptDetails' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'payload' on Entity 'Product' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'productBeforeEditJson' on Entity 'Product' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'payload' on Entity 'ZReport2' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'backendLayout' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'baseURL' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'baseURLs' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'cardPaymentConfig' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'giftCardActivationJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'giftCardSettingsJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'invoiceConfigurationJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'invoiceCustomersPayload' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'invoiceSettingsJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'libraryEventLogUuid' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'openCashRegisterJSON' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'payload' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'remoteAppEventsJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'sessionCacheJSON' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'suggestedOptionsJson' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
CoreData: warning: Property 'tokensForDisplayedInformation' on Entity 'StoreAccount' is using nil or an insecure NSValueTransformer.  Please switch to using "NSSecureUnarchiveFromData" or a subclass of NSSecureUnarchiveFromDataTransformer instead.
2019-11-13 13:05:44.737984+0100 iZettle SDK Sample[392:13850] [Process] kill() returned unexpected error 1
2019-11-13 13:05:44.790238+0100 iZettle SDK Sample[392:13850] [Process] kill() returned unexpected error 1
2019-11-13 13:05:44.791614+0100 iZettle SDK Sample[392:13850] [Process] kill() returned unexpected error 1
2019-11-13 13:05:44.791782+0100 iZettle SDK Sample[392:13850] [Process] kill() returned unexpected error 1
2019-11-13 13:05:44.791809+0100 iZettle SDK Sample[392:13850] [Process] kill() returned unexpected error 1
2019-11-13 13:05:44.903426+0100 iZettle SDK Sample[392:13850] [ProcessSuspension] 0x10bce41e0 - ProcessAssertion::processAssertionWasInvalidated()
2019-11-13 13:05:47.050704+0100 iZettle SDK Sample[392:14035] Task <DC1184C4-9729-4605-8892-7CD9D17F6EF9>.<1> finished with error [-999] Error Domain=NSURLErrorDomain Code=-999 "cancelled" UserInfo={NSErrorFailingURLStringKey=https://secure.izettle.com/api/resources/mobile/revisit, NSLocalizedDescription=cancelled, NSErrorFailingURLKey=https://secure.izettle.com/api/resources/mobile/revisit}
2019-11-13 13:05:47.053906+0100 iZettle SDK Sample[392:14030] Task <DC1184C4-9729-4605-8892-7CD9D17F6EF9>.<1> HTTP load failed, 0/0 bytes (error code: -999 [1:89])
2019-11-13 13:05:59.047878+0100 iZettle SDK Sample[392:13850] [TableView] Warning once only: UITableView was told to layout its visible cells and other contents without being in the view hierarchy (the table view or one of its superviews has not been added to a window). This may cause bugs by forcing views inside the table view to load and perform layout without accurate information (e.g. table view bounds, trait collection, layout margins, safe area insets, etc), and will also cause unnecessary performance overhead due to extra layout passes. Make a symbolic breakpoint at UITableViewAlertForLayoutOutsideViewHierarchy to catch this in the debugger and see what caused this to occur, so you can avoid this action altogether if possible, or defer it until the table view has been added to a window. Table view: <UITableView: 0x1028a9000; frame = (0 0; 414 736); clipsToBounds = YES; autoresize = RM+BM; gestureRecognizers = <NSArray: 0x281feb120>; layer = <CALayer: 0x28116e7e0>; contentOffset: {0, -12}; contentSize: {414, 0}; adjustedContentInset: {12, 0, 80, 0}; dataSource: <IZAccountListController: 0x101d501c0>>
2019-11-13 13:05:59.058665+0100 iZettle SDK Sample[392:13850] Unbalanced calls to begin/end appearance transitions for <IZAccountListController: 0x101d501c0>.
2019-11-13 13:06:14.795702+0100 iZettle SDK Sample[392:13850] Could not signal service com.apple.WebKit.Networking: 113: Could not find specified service`

I am using an iPhone 7 with ios 13.2.2

**Comment by saidsikira** [comment]
Hey @flawlessBadger, can you check if you're using the correct API key for your integration? Setting incorrect one could result in the issue you're seeing.

Thanks,
Said

**Comment by flawlessBadger** [comment]
That was the problem, thought I could test without an API key.

**Comment by bashleigh** [comment]
Hi @saidsikira I don't suppose there's any other issues that could cause this? I've double checked both my clientID and bundleID. Am I missing a step such as enable a particular URL etc? I have added my app scheme url in url types so it's not related to that. 

I'm using 
- swift 5
- xcode 12.4
- iso 14.2 (on iphone Xs)
- Flutter 1.22.5 and dart 2.10.4 (I know these are irelevant, worth noting if anything!)
Any other info you need or would like? 

Failing any other knowledege you may have! How can I debug this?

Looks like my exception is being throw from `iZettleSDK.shared().start(with: authenticationProvider)`.


---------------------------

Also worth mentioning; I can get to this point 

```
print("finished init")

iZettleSDK.shared().start(with: authenticationProvider)
print("finished start") <- this method is called
```

### Last logs 

```
23:01:45.879: [Debug] <IZLogoutService.m:38> Will log out account: <IZLogoutService: 0x283fc0ea0>
finished start
2021-02-22 23:01:45.950450+0000 Runner[6917:2388147] flutter: null //// This is the return from the invoke method
2021-02-22 23:02:16.005393+0000 Runner[6917:2387949] Could not signal service com.apple.WebKit.Networking: 113: Could not find specified service
```

**Comment by stagor** [comment]
Hi @bashleigh 
From what you are describing it seems as if there are something missing in the Info.plist file, please make sure that all the keys described [here](https://github.com/iZettle/sdk-ios#usage) are added to Info.plist.

**Comment by bashleigh** [comment]
@stagor ahhh yes, thanks I'll double check that. I did add one or 2 manually so I might've done it incorrectly! Thanks

**Comment by bashleigh** [comment]
The only step I've missed for now is contacting izettle abouting deploying the app and apply to the MFi program? Don't suppose that's a requirement for development? Not sure what else to try at this point?

**Comment by bashleigh** [comment]
@stagor any idea what else I might've missed?

**Comment by stagor** [comment]
@bashleigh It looks good, the only thing that seems to be missing, or is not visible in the screenshot is `URL Types`.
<img width="682" alt="Screenshot 2021-02-25 at 09 01 12" src="https://user-images.githubusercontent.com/57406220/109122291-a8e71800-7748-11eb-9581-7e7250bef17f.png">
Where `app-scheme` should be the same that you used in developer porta for `callback URL` and the same you use in:
``` Swift
iZettleSDKAuthorization(clientID: "...", callbackURL: ..., enforcedUserAccount: nil)
```

If you still experience problems - it's best to create a new issue and provide all the details there instead of discussing in this old closed one.

**Comment by bashleigh** [comment]
@stagor hey! Thanks for replying! Does the URL type have to start with the prefix `app-scheme` or can it be anything?

**Comment by stagor** [comment]
The app scheme can be anything, the callback url should look something like this: `may-awesome-app://login.callback` then the URL scheme is `may-awesome-app`

**Comment by bashleigh** [comment]
@stagor Hey! Just thought I'd let you know I resolved my issue! I'm not familiar with iOS swift etc so I presumed my catch block was catching an exception from the izettle start function and the exception was `Could not signal service com.apple.WebKit.Networking: 113: Could not find specified service`. Turns out this is not an issue, if I call the chargeAmount function, everything worked perfectly 👍  Thanks for your help!
